### ISSUE TITLE
firmware: GUI UX improvements

### ISSUE THREAD

**Issue by stilez** [opnsense/core#2579]
I'm not sure what exactly the issue here is, but something's clearly not as expected/not happy.

I've run through the check updates/update/unlock upgrades/upgrade cycle about 4 times now, and it just isn't doing as expected. For example, I just unlocked and upgraded from 18.1; after the upgrade and reboot I checked again and it says it's upgraded to 18.7-r1, but now has a second upgrade to 18.7-r2. (see screenshot). In any case, surely when it's asked to do an upgrade, it should jump to the latest version?

This is what I have from the GUI.  

- Under firmware settings, I'm on the production track. I've done an upgrade today, which should have got me to the latest production version. I did a search for updates after it rebooted. 
- The check is horribly self-conflicting.  
- It confirms I'm on the production track, but then says that immediately after an upgrading, I'm on an end of life release. 
- It says that the latest production version is 18.7-r2 (but **_also_** says the next major version is 18.7-r1 ?!).
- It says that the current in-use versions are opnsense NA (?!)  and opnsense-devel 18.7_r10 (but I'm not on devel), and that the "action needed" is that devel must be upgraded and production action is "new" (whatever that means).

**Screenshot #1:**
![opnsense](https://user-images.githubusercontent.com/3777403/43327352-b8635c52-91aa-11e8-9b6b-6f003dff3f17.png)

**Screenshot #2:**
![opnsense2](https://user-images.githubusercontent.com/3777403/43327354-b8853264-91aa-11e8-8e57-166c8763c62c.png)

Basically it seems horribly confused, and whatever I do, I can't seem to get it to a "you are up to date" position. Maybe it should have a filter "show only updates for my current track (_[production | devel ]_)" vs. "Show updates for all tracks"? Or only show updates relevant to develop/production, depending which you've selected, and not try to show both? And what's with the "unlock" thing anyway? What function does that serve?

**Comment by L1ghtn1ng** [comment]
I have run into this as well on my VM
________________________________

**Comment by fichtner** [comment]
It‘s the dev version. It loops major changes pinned to 18.7.r1. It’s confusing, but it’s for testing. Basically the blue box are minor updates and the yellow box are major upgrades. This doesn’t happen on 18.1 when the time for 18.7 comes.

**Comment by stilez** [comment]
Any objection if I try to make some time in thr next while, to hit this one hard with the pull request hammer, so as to try to make it more intuitive or behave more sensibly? :)

**Comment by fichtner** [comment]
The underlying mechanisms are fixed around what pkg and our self-written kernel/base update does. The UX certainly isn't perfect so there's room to improve that so feel free to research or discuss certain aspects. :)

For clarity, let's answer your individual points:

> Under firmware settings, I'm on the production track. I've done an upgrade today, which should have got me to the latest production version. I did a search for updates after it rebooted.

The production track does not upgrade to 18.7 yet. So you need to go to the dev version which offers a fixed path from 18.1 to 18.7.r1.

> The check is horribly self-conflicting.

Not necessarily. You click check and it offers you updates (if applicable) and upgrades (if applicable). These work with fundamentally different workflows in the backend.

> It says that the latest production version is 18.7-r2 (but also says the next major version is 18.7-r1 ?!).

Yes, again, major upgrades are fixed to a version. We did not add 18.7.r2 into this upgrade path. You just need to cross over from 18.1 to 18.7 once, from there you can get the normal updates.

> It says that the current in-use versions are opnsense NA (?!) and opnsense-devel 18.7_r10 (but I'm not on devel), and that the "action needed" is that devel must be upgraded and production action is "new" (whatever that means).

18.7.r_10 is the development version of 18.7.r1. It would not make sense to have both the development version and "production" version to have the same 18.7.r1 version tag. All our update version numbers follow the production version of the core package, even though there are development packages, plugin packages of both varieties and third party packages with different version numbers.

> Screenshot 1

Click "Update now". The yellow upgrade window is only there because the dev version loops the major upgrade for testing. This will not happen in production.

Your core package is "opnsense-devel" so it needs to be replaced with "opnsense" when you go from development back to production. "opnsense" shows as N/A because it's not installed. This is a pkg mechanic. The version is 18.7.r2 because 18.7.r1 is no longer on the update path (even though it's still used for the upgrades).

Basically, all updates use pkg's online repository. All upgrades use archived sets to do a full offline update.

> Screenshot 2

Same as 1, sometimes "pkg" refuses to give a hint about packages that are going to be deinstalled. We have no control over this.

**Comment by fichtner** [comment]
It also needs to be pointed out that the development version uses the master branch with is both version 18.1 and 18.7 at the same time. So while in 18.1 you expect the upgrade to advertise itself in 18.7 it does too even though you're already there. This does not happen in stable/18.1 or stable/18.7 which is used by the production type.

**Comment by fichtner** [comment]
@stilez hi there, did you make any progress? what we discussed makes sense each time I see the firmware page, but I'd rather not start without you so let me know if/when you find the time to work on it. cheers :)

**Comment by stilez** [comment]
I didn't start yet, mainly because what is clear to me is, until there is a clear idea what you (and anyone else interested) think it should look like, and how it should function, there is no point taking a unilateral guess :) Maybe, can you do.a quick mock up and/or describe what you think it should look like and how it should work, and email it (to keep this thread short for others). Then I can understand your ideas, and come back to you on it, hopefully quickly reaching a consensus to post here, which is more likely to not need revision later on.

As an aside, it will be my 1st time working on OPNsense's mvc setup;  I might need to ask a bit more "dummy" questions by email first time round, or watch if you do something  not sure about, if any things aren't clear to me from the docs, dev guide or codebase. It should be a "once off" and "easy questions" to orient myself and get it right - once I've understood , I won't need to ask again on other PRs. Crossing fingers that this will be OK :)

**Comment by fichtner** [comment]
sure, happy to answer all questions :)

**Comment by fichtner** [comment]
A few things changed a bit over a year ago, but all in all we have no questions about the page layout or a lack of clarity.
